### PR TITLE
chore: make CREATE OR REPLACE use execution step and pseudocolumn versions of original query

### PR DESCRIPTION
**Pull request by Sullivan-Patrick**
### Description 
As part of [KLIP-50](https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-50-partition-and-offset-in-ksqldb.md), When using `CREATE OR REPLACE` to upgrade a query that has an old pseudocolumn version number, we want to use the old one to ensure backwards compatibility.

In the case of `TableSourceV1`s that are being upgraded from, we want to continue to use the V1 step as to not change the KStreams topology. For context on why the V2 source exists and how it changes the topology, see https://github.com/confluentinc/ksql/pull/7990, https://github.com/confluentinc/ksql/pull/8072, https://github.com/confluentinc/ksql/pull/8111.

This PR makes it so replaced queries use old `SourceStep.pseudoColumnVersion` fields, and queries with a `TableSourceV1` source continue to do so.

### Testing done 
Unit tests in `SchemaKSourceFactoryTest`, SQL testing tool tests. It's worth noting that the SQL testing tool isn't fully compatible with `ROWPARTITION` and `ROWOFFSET` currently. These fields are able to be selected as part of a query, but they do not work directly with `ASSERT`, and when checked indirectly (by selecting them in a new query and using `ASSERT` on the new stream's aliased columns), the values are always 0. Regardless, these tests to add some extra coverage, so I think it's good to have them.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-engine/src/main/java/io/confluent/ksql/structured/SchemaKSourceFactory.java
Comment (vcrfxia) [inline ksqldb-engine/src/main/java/io/confluent/ksql/structured/SchemaKSourceFactory.java:202]: nit: I don't think this comment is necessary. If `KsqlConfig.KSQL_ROWPARTITION_ROWOFFSET_ENABLED` is deleted, then this line will have to be updated (else compilation will fail) and it's pretty clear how this should be updated when that happens.

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryEngine.java
Comment (vcrfxia) [inline ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryEngine.java:77]: nit:
```suggestion
      final Optional<PlanInfo> oldPlanInfo
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/physical/PhysicalPlanBuilder.java
Comment (vcrfxia) [inline ksqldb-engine/src/main/java/io/confluent/ksql/physical/PhysicalPlanBuilder.java:60]:
```suggestion
      final Optional<PlanInfo> oldPlanInfo
```

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/PlanInfo.java
Comment (vcrfxia) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/PlanInfo.java:74]: nit:
```suggestion
  public Set<ExecutionStep<?>> getSources() {
```

File: ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java:326]: Looks like there's no test checking that TableSourceV2 with an existing query uses the existing pseudocolumn version? Unless it throws an error to use an old version with TableSourceV2, I think it makes sense to add the test even if it's unrealistic at this time (you can leave a comment to point out the fact that it's unrealistic, if that makes it feel less weird).

File: ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java:381]:
```suggestion
  public void shouldReplaceWindowedTableSourceWithMatchingPseudoColumnVersion() {
```

File: ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java:476]: Why is this method always using `streamSource` even from the tests that are testing tables, etc? I think we should make this wiring more airtight -- you can pass in the source step to mock as a param if you don't want to make copies of this method.

File: ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java:72]: Some of the tests added in this file fail unless `OLD_PSEUDOCOLUMN_VERSION` is exactly `SystemColumns.LEGACY_PSEUDOCOLUMN_VERSION_NUMBER`. In other words, it doesn't work for this test file to use a generic "old" version and a generic "new version". As such, I think the tests should just use `LEGACY_PSEUDOCOLUMN_VERSION_NUMBER` directly, instead of defining an alias. Similarly, we can just use `SystemColumns.CURRENT_PSEUDOCOLUMN_VERSION_NUMBER` in place of defining a new variable `NEW_PSEUDOCOLUMN_VERSION` as well.

**Review by vcrfxia — COMMENTED** [review]
Thanks @Sullivan-Patrick . LGTM overall assuming this has been tested. Did you get a chance to look at KsqlTesterTest? Also, we need to test this with CREATE SOURCE TABLE. I'm not familiar enough with the implementation of CREATE SOURCE TABLE to comment on whether this PR impacts it or not. Let's be sure to follow up on that before merging.

File: ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java
Comment (Sullivan-Patrick) [reply ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java:381]: Why this but not the others with the same convention (ie `shouldReplaceOldWindowedStreamSourceWithMatchingPseudoColumnVersion`)? I don't feel strongly that the `Old` is necessary in any, though.

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/PlanInfo.java
Comment (vcrfxia) [reply ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/PlanInfo.java:74]: `Sources` is a private class, so I don't think the name is ambiguous. If you prefer we can do `getSourceSteps()` instead, but I do think `getSourceSet()` is no good since the fact that it's a set is not relevant (and can be seen at a glance from the type itself).

File: ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java
Comment (vcrfxia) [reply ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java]: I'm all for updating the others too! 😁

Comment (vcrfxia) [reply ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java]: I think "old" is unnecessary/confusing because:

* there's no such thing as an "old" windowed table source, i.e., there's only WindowedTableSourceV1 and no WindowedTableSourceV2
* if "old" is meant to mean "existing" rather than "of an old version", then it is redundant because if it's being replaced then it must be old
* table sources actually do have two versions (TableSourceV1 and TableSourceV2) so "old" does make sense in this context (though ideally we'd say TableSourceV1 and TableSourceV2 rather than OldTableSource and NewTableSource, in order to age better). Using "old" in places where it doesn't make sense detracts from places where it does make sense.

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PlanBuildContext.java
Comment (vcrfxia) [inline ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PlanBuildContext.java:82]:
```suggestion
        planInfo
```

File: ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java:258]: Why are we passing in a windowedStreamSource for a test that tests non-windowed stream sources?

File: ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java:269]: nit: you can use a static important to shorten appearances of `SystemColumns.LEGACY_PSEUDOCOLUMN_VERSION_NUMBER` to simply `LEGACY_PSEUDOCOLUMN_VERSION_NUMBER`, and similarly for `CURRENT_PSEUDOCOLUMN_VERSION_NUMBER`.

File: ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java:348]: What is an "OldWTableSourceV2"? See other discussion about the meaning of "old" in these tests.

File: ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/structured/SchemaKSourceFactoryTest.java:346]: Clarify what this comment means -- it's only true today because TableSourceV2 was introduced alongside pseudocolumn versioning, and there are only two pseudocolumn versions today. This test may become relevant in the future (else there's no point adding it).

File: ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql:2]:
```suggestion
--@test: upgrade stream across pseudocolumn versions
```

File: ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql:4]: Why does this need to be set? Isn't this already the default?

File: ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql:25]: This test doesn't seem to be testing what it claims to test? As written, the test verifies that CSAS is allowed on sources that have been CREATE OR REPLACE'd. It does not verify that CREATE OR REPLACE is allowed/disallowed for ROWPARTITION and ROWOFFSET. I think the more useful tests to add in this file are to check that the user gets a good error message if they try to CREATE OR REPLACE an old query and add ROWPARTITION or ROWOFFSET.

**Review by vcrfxia — COMMENTED** [review]
Thanks @Sullivan-Patrick -- some comments inline. Thanks for looking into KsqlTesterTest behavior with regards to ROWOFFSET/ROWPARTITION. Can you file a Github issue for the lack of behavior, so we don't lose track? Also, did you get a chance to try CREATE OR REPLACE with CREATE SOURCE TABLE, to check that the changes here haven't broken existing behavior there?

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/PlanInfo.java
Comment (Sullivan-Patrick) [reply ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/PlanInfo.java:74]: I think the private class justification is a good one, and `getSources()` is concise.

File: ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql
Comment (Sullivan-Patrick) [reply ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql:25]: Good point- this was just  a sanity check that we are able to actually use ROWPARTITION//ROWOFFSET (ie the feature was turned on). That doesn't really seem great, as the test is testing itself a bit (the SET functionality), and nothing else. The test you mentioned is a lot better.

File: ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql:2]: I thought we discussed removing this test (and the one below) in https://github.com/confluentinc/ksql/pull/8097#discussion_r717135179?

File: ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql:79]: I don't think we need all four of these tests (each of ROWPARTITION/ROWOFFSET in each of SELECT/WHERE clauses). Let's just keep two of them: 
- rename `should fail on CREATE OR REPLACE of an old query with ROWPARTITION in SELECT clause` to `should fail on CREATE OR REPLACE of a legacy query with ROWPARTITION`
- rename `should fail on CREATE OR REPLACE of a legacy query with ROWOFFSET in WHERE clause` to `should fail on CREATE OR REPLACE of a legacy query with ROWOFFSET`

File: ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql:62]: This error message is not great. Ideally we'd print a more targeted error to explain that new pseudocolumns are not supported in CREATE OR REPLACE on existing queries.

We can achieve this by updating `CodeGenRunner#addRequiredColumn()` where this error message is thrown to add a check for whether the column name matches a pseudocolumn name and add to the error message something like "If this is a CREATE OR REPLACE query, pseudocolumns added in newer versions of ksqlDB after the original query was issued are not available for use in CREATE OR REPLACE."

**Review by vcrfxia — COMMENTED** [review]
Thanks @Sullivan-Patrick -- LGTM besides the comments inline.

File: ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql
Comment (Sullivan-Patrick) [reply ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql:2]: I interpreted this as adding tests, not necessarily removing. I'll go ahead and do that.

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenRunner.java
Comment (vcrfxia) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenRunner.java:325]: Sorry if my previous message wasn't clear -- this additional message should be added in addition to the original message. The original message is still important for users to know, otherwise they don't know what the actual failure was.

**Review by vcrfxia — COMMENTED** [review]
Thanks @Sullivan-Patrick -- almost there! One comment inline.

File: ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql
Comment (Sullivan-Patrick) [reply ksqldb-functional-tests/src/test/resources/sql-tests/query-upgrades/rowpartition-rowoffset.sql:4]: I think you were clear enough, I'm being a bit sloppy right now as I'm scrambling to tie up all the loose ends before end of week.

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenRunner.java
Comment (Sullivan-Patrick) [reply ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenRunner.java]: I think you were clear enough, I'm being a bit sloppy right now as I'm scrambling to tie up all the loose ends before end of week.

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenRunner.java
Comment (vcrfxia) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenRunner.java:322]: nit: let's only compute this if we need to the throw the error message. You can define a helper method `fieldNotFoundErrorMessage(columnName)` to return the appropriate error message string. Besides being more efficient, this will also help with readability.

**Review by vcrfxia — COMMENTED** [review]
Thanks @Sullivan-Patrick -- LGTM besides the minor comment inline, and the failing build. Let's merge afterwards! 🎉

**Review by vcrfxia — APPROVED** [review]

**Pull request merged**
